src/test/java/net/minecraftforge/debug/world/BiomeModifierTest.java

```diff
@@ -89,86 +88,49 @@ public BiomeModifierTest()
         serializers.register(modBus);
         serializers.register(TEST, TestModifier::makeCodec);
 
-        // Biome modifiers don't need to be registered when defined in json, but they do need to be registered if we are to datagenerate the jsons.
-        // We'll also datagenerate a placedfeature json (using our own placedfeature avoids feature cycle problems when we add it to biomes).
-        final DeferredRegister<PlacedFeature> placedFeatures = DeferredRegister.create(Registry.PLACED_FEATURE_REGISTRY, MODID);
-        placedFeatures.register(modBus);
-        placedFeatures.register(BASALT_PILLARS,
-            () -> new PlacedFeature((Holder<ConfiguredFeature<?,?>>) (Holder<? extends ConfiguredFeature<?,?>>)NetherFeatures.LARGE_BASALT_COLUMNS,
-                List.of(CountOnEveryLayerPlacement.of(1), BiomeFilter.biome())));
-
         modBus.addListener(this::onGatherData);
     }
 
     private void onGatherData(GatherDataEvent event)
     {
         // Example of how to datagen datapack registry objects.
-        // Datapack registry objects referred to by other datapack registry objects must be registered first.
-        DataGenerator generator = event.getGenerator();
-        final Path outputFolder = generator.getOutputFolder();
-        final RegistryOps<JsonElement> ops = RegistryOps.create(JsonOps.INSTANCE, RegistryAccess.BUILTIN.get());
-        final Gson gson = new GsonBuilder().setPrettyPrinting().create();
-        final String directory = PackType.SERVER_DATA.getDirectory();
-        Registry<PlacedFeature> placedFeatures = ops.registry(Registry.PLACED_FEATURE_REGISTRY).get();
-
-        // prepare to datagenerate our placed feature
-        final String featurePathString = String.join("/", directory, MODID, Registry.PLACED_FEATURE_REGISTRY.location().getPath(), BASALT_PILLARS + ".json");
-        final Path featurePath = outputFolder.resolve(featurePathString);
-        final PlacedFeature feature = placedFeatures.get(BASALT_PILLARS_RL);
-
-        // prepare to datagenerate our biome modifier
-        final ResourceLocation biomeModifiersRegistryID = ForgeRegistries.Keys.BIOME_MODIFIERS.location();
-        final String biomeModifierPathString = String.join("/", directory, MODID, biomeModifiersRegistryID.getNamespace(), biomeModifiersRegistryID.getPath(), MODIFY_BADLANDS + ".json");
-        final Path biomeModifierPath = outputFolder.resolve(biomeModifierPathString);
+        final DataGenerator generator = event.getGenerator();
+        final ExistingFileHelper existingFileHelper = event.getExistingFileHelper();
+        
+        // Any reference holders our objects have must come from this same RegistryAccess instance,
+        // or encoding our objects will fail.
+        final RegistryAccess registries = RegistryAccess.builtinCopy();
```

Comment:
It is my understanding that making a builtin copy is slow. If every mod is doing this for datagen, that seems a bit unnecessarily slow. Perhaps Forge should store a memoized builtinCopy() that just isn't frozen, for datagen purposes.

Reply:
Since this is datagen where only a single mod or at least a handfull of mods is running, this is not a problem here.